src/utils/PathUtils.ts

```diff
@@ -0,0 +1,127 @@
+import { PathPart } from '../fhirtypes';
+import { splitOnPathPeriods } from '../fhirtypes/common';
+import { CaretValueRule, Rule } from '../fshtypes/rules';
+
+/**
+ * Parses a FSH Path into a more easily usable form
+ * @param {string} fshPath - A syntactically valid path in FSH
+ * @returns {PathPart[]} an array of PathParts that is easier to work with
+ */
+export function parseFSHPath(fshPath: string): PathPart[] {
+  const pathParts: PathPart[] = [];
+  const splitPath = fshPath === '.' ? [fshPath] : splitOnPathPeriods(fshPath);
+  for (const pathPart of splitPath) {
+    const splitPathPart = pathPart.split('[');
+    if (splitPathPart.length === 1 || pathPart.endsWith('[x]')) {
+      // There are no brackets, or the brackets are for a choice, so just push on the name
+      pathParts.push({ base: pathPart });
+    } else {
+      // We have brackets, let's  save the bracket info
+      let fhirPathBase = splitPathPart[0];
+      // Get the bracket elements and slice off the trailing ']'
+      let brackets = splitPathPart.slice(1).map(s => s.slice(0, -1));
+      // Get rid of any remaining [x] elements in the brackets
+      if (brackets[0] === 'x') {
+        fhirPathBase += '[x]';
+        brackets = brackets.slice(1);
+      }
+      pathParts.push({ base: fhirPathBase, brackets: brackets });
+    }
+  }
+  return pathParts;
+}
+
+/**
+ * Assembles a PathPart array back to its original form
+ * @param {PathPart[]} pathParts - An array of pathParts
+ * @returns {string} path - The path corresponding to those pathParts
+ */
+export function assembleFSHPath(pathParts: PathPart[]): string {
+  let path = '';
+  pathParts.forEach((pathPart, i) => {
+    path += pathPart.base;
+    pathPart.brackets?.forEach(bracket => (path += `[${bracket}]`));
+    if (i < pathParts.length - 1) {
+      path += '.';
+    }
+  });
+  return path;
+}
+
+/**
+ * Replaces soft indexs in rule paths with corresponding numbers
+ * @param {Rule[]} rules - An array of Rules
+ */
+export function resolveSoftIndexing(rules: Rule[]): void {
+  const pathMap: Map<string, number> = new Map();
+  const caretPathMap: Map<string, Map<string, number>> = new Map();
+
+  // Parsing and separating rules by base name and bracket indexes
+  const parsedRules = rules.map(rule => {
+    const parsedPath: any = { path: parseFSHPath(rule.path) };
```

Comment:
I think you can actually type this more specifically than `any`:
```ts
    const parsedPath: { path: PathPart[], caretPath?: PathPart[] } = { path: parseFSHPath(rule.path) };
```
Then you'll get some of the type-safety features of having a strongly typed object.

Reply:
Oh, I see.  You're actually re-assigning these as strings later on.... In that case you could still add `| string` to the type info for each.  But I'm thinking maybe we can avoid that re-assignment so that this can stay as a single static type.  More in a later comment...